workspace/src/localization/localization_py/localization_py/particle_filter.py

```diff
@@ -0,0 +1,219 @@
+import numpy as np
+import math
+import matplotlib.pyplot as plt
+import random
+from scipy.stats import wasserstein_distance
+import yaml
+
+class particle_filter:
+    """
+    A basic Particle Filter implementation.
+    
+    This currently uses the 4DOF motion model. The Motion Model inputs are throttle and steering, and the observation model uses GPS and Magnetometer as sensors. States are cartesian coordinates x, y, heading theta, and velocity v. Further reference can be found at: https://sbel.wisc.edu/wp-content/uploads/sites/569/2023/07/TR-2023-05.pdf.
+
+    Attributes:
+        c_1: The motor resistence torque linear component.
+        c_0: The motor resistence torque constant component.
+        l: The length of the vehicle.
+        r_wheel: The radius of the wheel.
+        i_wheel: the intertia of the wheel.
+        gamma: The gear ratio.
+        tau_0: The stalling torque for the motor torque model.
+        omega_0: The maximum no-load speed for the motor torque model.
+        df_1_dv: The derivative of the scaled motor torque relative to the velocity of the vehicle.
+        Q: The covariane matrix for our states.
+        R: The covariance matrix for our sensors.
+        P: The predicted covariance estimate.
+        show_animation: Boolean for whether or not to show the particle animation.
+        dt: The timestep for the filter.
+        num_particles: The number of particles maintained by the filter.
+        particles: The array of particles.
+        hpx: The array of x coordinate histories for each particle. Used for the animation.
+        hpy: The array of y coordinate histories for each particle. Used for the animation.
+        particle_distr_dist: The expected noise distribution for the position of the particles.
+        particle_distr_head: The expected noise distribution for the heading of the particles.
+        timestep: the current timestep. Used for the animation.
+    
+    """
+
+    def __init__(self, dt):
+        """Initialize the PF.
+
+        Initialize each of the global variables for the PF. Load the dynamics parameters for the vehicle from an external YAML file.
+
+        Args:
+            dt: The timestep at which this filter will operate at. 
+        """
+        with open('/home/art/art/workspace/src/localization/localization_py/localization_py/4DOF_dynamics.yml', 'r') as yaml_file:
```

Comment:
hardcode again. also is this duplicate from EKF.py?

Reply:
Yes - maybe better to have the dynamics model as a separate class? I think that that would work better for Huzaifa as well...

Reply:
I took a look at this and I think this is much better than what it was before for mainly 1 reason. 

The state estimator "uses" a dynamics model and so it is appropriate to separate these two entities into different classes. A state estimator class should not implicitly define a dynamics model (the way it was before) as I should be able to provide any dynamics model, provided it contains certain member functions (here "motion_model" and "calc_F") to a instance of the state estimator class without having to modify the class. 

That being said, there should be an exception raised for when a dynamics class without these two methods is provided as a member.  
Or maybe a Dynamics abstract class that has the 2 required methods - users would then have to create child Dynamics classes that overload these two methods.